File: algorithms/2_medium/D/01381_design_a_stack_with_increment_operation/solution.cpp
class CustomStack {
    vector<int> stack;
    const size_t max_size;
public:
    CustomStack(int maxSize) :  max_size{static_cast<size_t>(maxSize)} { }
    
    void push(int x) {
        if (stack.size() < max_size) {
            stack.push_back(x);
        }
    }
    
    int pop() {
        if (stack.empty()) {
            return -1;
        }

        int res{stack.back()};
        stack.pop_back();
        return res;
    }
    
    void increment(int k, int val) {
        const size_t end_index{min(stack.size(), static_cast<size_t>(k))};

        for (size_t i{0}; i < end_index; ++i) {
            stack[i] += val;
        }
    }
};

/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack* obj = new CustomStack(maxSize);
 * obj->push(x);
 * int param_2 = obj->pop();
 * obj->increment(k,val);
 */

File: algorithms/2_medium/D/01381_design_a_stack_with_increment_operation/solution.py
class CustomStack:

    def __init__(self, maxSize: int):
        self.max_size = maxSize
        self.stack = [0] * maxSize
        self.current_head = 0

    def push(self, x: int) -> None:
        if self.current_head < self.max_size:
            self.stack[self.current_head] = x
            self.current_head += 1

    def pop(self) -> int:
        if self.current_head == 0:
            return -1
        self.current_head -= 1
        return self.stack[self.current_head]

    def increment(self, k: int, val: int) -> None:
        for i in range(0, min(self.current_head, k)):
            self.stack[i] += val


# Your CustomStack object will be instantiated and called as such:
# obj = CustomStack(maxSize)
# obj.push(x)
# param_2 = obj.pop()
# obj.increment(k,val)
